internal/credentials/kubernetes/github/app.go

```diff
@@ -69,17 +75,51 @@ func (p *AppCredentialProvider) Supports(
 		strings.TrimSpace(string(data[installationIDKey])) != "" && strings.TrimSpace(string(data[privateKeyKey])) != ""
 }
 
+// GetCredentials implements the credentials.Provider interface for GitHub Apps.
+// It returns GitHub installation access tokens scoped to a repository, with
+// optional restrictions enforced from the `project-repos` annotation.
 func (p *AppCredentialProvider) GetCredentials(
-	_ context.Context,
-	_ string,
+	ctx context.Context,
+	project string,
 	credType credentials.Type,
 	repoURL string,
 	data map[string][]byte,
 ) (*credentials.Credentials, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues()
 	if !p.Supports(credType, repoURL, data) {
 		return nil, nil
 	}
 
+	// Extract the project-repos JSON from annotation
+	// in data or directly from data key
+	projectReposJSON, hasProjectRepos := data[kargoapi.AnnotationProjectReposKey]
+
+	var (
+		allowedRepos []string
+		restricted   bool
+	)
+
+	if hasProjectRepos && len(projectReposJSON) > 0 {
+		var projectRepoMap map[string][]string
+		if err := json.Unmarshal(projectReposJSON, &projectRepoMap); err != nil {
+			// If JSON is invalid, we mark as restricted (deny by default)
+			logger.Debug("error unmarshalling project-repos JSON")
+			restricted = true
```

Comment:
@krancour shouldn't we return an error here instead of denying by default?

Reply:
Here, yes. (If this somehow came from one of my suggested code edits, I apologize.  Editing on GH is prone to mistakes.) In other spots, I've mentioned that not finding a credit isn't the same as an error. But this is a case where we can guess that there was _supposed_ to be a policy here telling us something important, but we can't guess what it was. Moving on to the next provider isn't the right thing to do. The right thing is to surface the error and let the user straighten it out.